flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherImpl.java

```diff
@@ -98,18 +120,26 @@ private ChannelStateCheckpointWriter buildWriter(CheckpointStartRequest request)
                 streamFactoryResolver.resolveCheckpointStorageLocation(
                         request.getCheckpointId(), request.getLocationReference()),
                 serializer,
-                () -> writers.remove(request.getCheckpointId()));
+                () -> {
+                    checkState(
+                            request.getCheckpointId() == ongoingCheckpointId,
+                            "The ongoingCheckpointId[%s] was changed when clear writer of checkpoint[%s], it might be a bug.",
+                            ongoingCheckpointId,
+                            request.getCheckpointId());
+                    this.writer = null;
+                });
```

Comment:
I think this `checkState` might be working now, only because we do not fail/reset writer if checkpoint is aborted. Otherwise this would fail when there is a race condition between
1. receiving last of the checkpoint barriers from the input `ChannelStateWriter#finishInput` for a checkpoint `42` in `CheckpointBarrierHandler`
2. aborting checkpoint `43` for example.


Maybe the best equivalent of the old code would be something like:
```
if (request.getCheckpointId() < ongoingCheckpointId) {
  // ignore obsolete request, do nothing
} else if (request.getCheckpointId() == ongoingCheckpointId) {
  this.writer = null;
} else { 
  throw new IllegalStateException("This should never happened, trying to completed checkpoint[%s] while the ongoingCheckpointId[%s] is older", ...);
}
```

Since the the code on the master also doesn't check if `writers.remove(42)` removed anything or not?

Reply:
Hi @pnowojski 

As I mentioned in the last comment, `writer` cannot be set to `null` during abort checkpoint.

> If the 2. happens after 1., this checkState would fail, right?

I don't think so. Actually, all `abort`, `start`, `writeInput`, `finishInput`, `writeOutput` or `finishOutput` are executed in the `Channel state writer Executor` Thread. It is single thread, so the race condition isn't exist.

Please correct me if anything is wrong.

Reply:
BTW, after our analysis, the code of the master branch has a memory leak. When checkpoint fails, `Map<Long, ChannelStateCheckpointWriter> writers;` cannot be cleaned up until the end of the Task.

After this PR, it will also be fixed.